gax-grpc/src/main/java/com/google/api/gax/grpc/OperationPollingCallable.java

```diff
@@ -0,0 +1,125 @@
+/*
+ * Copyright 2017, Google Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are
+ * met:
+ *
+ *     * Redistributions of source code must retain the above copyright
+ * notice, this list of conditions and the following disclaimer.
+ *     * Redistributions in binary form must reproduce the above
+ * copyright notice, this list of conditions and the following disclaimer
+ * in the documentation and/or other materials provided with the
+ * distribution.
+ *     * Neither the name of Google Inc. nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.google.api.gax.grpc;
+
+import com.google.api.core.ApiClock;
+import com.google.api.core.ApiFuture;
+import com.google.api.core.ApiFutures;
+import com.google.api.gax.retrying.ExponentialRetryAlgorithm;
+import com.google.api.gax.retrying.ResultRetryAlgorithm;
+import com.google.api.gax.retrying.RetrySettings;
+import com.google.api.gax.retrying.TimedAttemptSettings;
+import com.google.longrunning.GetOperationRequest;
+import com.google.longrunning.Operation;
+import com.google.longrunning.OperationsClient;
+import io.grpc.Status;
+import java.util.concurrent.ExecutionException;
+
+/**
+ * A future callable, which relies on callback chaining to execute server polling (asking for a
+ * status of a specific operation: in progress, completed, canceled etc.). This implementation is
+ * called from {@link AttemptCallable}, essentially using for polling the same logic which is used
+ * for retrying.
+ *
+ * @param <RequestT> type of the request
+ */
+class OperationPollingCallable<RequestT> implements FutureCallable<RequestT, Operation> {
+  private final OperationsClient operationsClient;
+  private final ApiFuture<Operation> initialFuture;
+
+  OperationPollingCallable(ApiFuture<Operation> initialFuture, OperationsClient operationsClient) {
```

Comment:
Yes. This class also does a lot of other crazy things (to me at least). It basically, together with AttemptCallable juggles with various futures in brutal manner. 

When I think about it it gives me headache. There are **tasks of different type**, representing some phase of the polling process, all executed by same threadpool, an yet, at any given point in time, there is only **one instance** of a task in the pool (either scheduled or running).

The specific reason, why this class accepts ApiFuture instead of FutureCallable: this allows to pass that future to returned OperationFuture object at the appropriate time. It is sill possible to avoid doing that, but it will involve more settable futures, listeners and weird circular dependencies and interactions between OperationFuture and a OperationPollingCallable. I really-really don't want to do that, especially, when there is a much simpler, more robust and faster solution. Also, this resembles how it was done in previous implementation. Plus this is a package-private class.

Also, originally I had it receiving FutureCallaable initialCallable in the constructor. But ended up changing that to support the OperationFuture case. This was also one of the reasons why I wanted to get rid from OperationFuture at all.